Add property API to InternetSocket

artokin opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Add new API to network interface.

Property API allows application to ask network specific properties
from the ~NetworkInterface~InternetSocket . Queried information can be used to avoid
network congestion by adjusting transmission jitter and retry
timeouts.

~Please note, this work is in progress.~

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@mikter 
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Commit:
Add getsockopt option to reading network property

Add getsockopt options NSAPI_LATENCY and NSAPI_STAGGER to read
network specific timing constraints from socket.
-NS_LATENCY returns estimated latency to given address.
-NSAPI_STAGGER returns estimated initial delay that application
 should wait before transmitting data to network.

Application can use the new options to avoid network congestion by
adjusting transmission delays and retry timeouts.

AnttiKauppila commented on features/netsocket/NetworkInterface.h:
```diff
@@ -378,6 +378,15 @@ class NetworkInterface: public DNS {
+     * @return NSAPI_ERROR_OK on success
+     * @return NSAPI_ERROR_UNSUPPORTED, NSAPI_ERROR_NO_CONNECTION
+     */
+    virtual nsapi_error_t get_interface_property(unsigned int data_amount, unsigned int *jitter);
```
Why is it needed in top level?

artokin replied:
Having API here allows application to be independent of actual interface.

kjbracey replied:
The ROM cost here due to the virtual methods, particularly if broken up, is significant. And these are quite specialised calls.

Therefore my inclination would be to do this via new reason codes for `getsockopt`, avoiding adding anything to the vtable.

Particularly as something like RTT information may in principle be connection/remote-peer-specific, not simply a property of the interface, and the application ultimately is going to want to be working with a socket anyway.

Note that you can still have "nice wrappers" around a setsockopt, such as `InternetSocket::join_multicast_group` or `TLSSocket::set_hostname` (offboard implementation). They would be non-virtual, so no overhead.

mikter commented on features/netsocket/InternetSocket.cpp:
```diff
@@ -114,6 +114,44 @@ int InternetSocket::leave_multicast_group(const SocketAddress &address)
+    ret = this->getsockopt(NSAPI_SOCKET, NSAPI_STAGGER, (void *)&nsapi_stagger, &opt_len);
+    if (ret == NSAPI_ERROR_OK) {
+        // success, stagger found
+        *stagger_min = nsapi_stagger.stagger_min;
```
Should you have null checks for parameter existence so this can be called just to get the rand_value only

artokin replied:
Absolutely! Added in the next commit.

Commit:
Review corrections

-Check pointer before writing to it. It allows application to ask
 only some of the  stagger values.
-Change type of data_amount in get_stagger_estimate_to_address to be
 uint16_t.

teetak01 reviewed (commented):
Should this not happen via NetworkInterface instead?

@yogpan01

teetak01 reviewed (dismissed):
Needs to be still reviewed further. We will get back to this tomorrow.

Commit:
Corrections from review meeting

-Rename get_latency_estimate_to_address, use rtt instead of latency
-Update descriptions of added methods

Commit:
Add unittests for added methods

Add unit tests for added methods:
 -get_rtt_estimate_to_address
 -get_stagger_estimate_to_address

kjbracey commented on features/netsocket/InternetSocket.cpp:
```diff
@@ -114,6 +114,54 @@ int InternetSocket::leave_multicast_group(const SocketAddress &address)
+    // Set up address
+    memcpy(ns_api_latency_req.addr, address.get_ip_bytes(), 16);
+
+    ret = this->getsockopt(NSAPI_SOCKET, NSAPI_LATENCY, (void *)&ns_api_latency_req, &opt_len);
```
Unnecessary cast.

kjbracey reviewed (dismissed):
Seems fine. Not 100% sure of the value of the wrapper around the socket option - stagger is a complex multivalue call, so not sure that all the pointer out parameters are really any simpler than using the socket option with the structure output. But I guess we set the precedent with the multicast options.

Commit:
Remove unnecessary casts from stagger/RTT API
